### PR TITLE
Create a new MmCommunicationDxe driver

### PR DESCRIPTION
**Pull request by td36**
# Description
StandaloneMmPkg: Add a new MmCommunicationDxe driver

Add a new MmCommunicationDxe driver. The driver is to:
1.Install gEfiMmCommunication2ProtocolGuid
2.Install gEfiMmCommunicationProtocolGuid
3.Create the notifications of some protocols and event that the Standalone Mm requires

## How This Was Tested

Tested in Intel internal platform

### CODE REVIEW

**Comment by td36** [comment]
Hi @jiaxinwu , @niruiyu , could you help to review the PR to create MmCommunicationDxe driver?

**Comment by mdkinney** [comment]
Please provide more background on why this new driver is required.

**Comment by mdkinney** [comment]
Some of the commit messages have more than one Signed-off-by.  There should only be one of those tags in each commit.  If more than one author contributed to the change, then additional Co-authored-by tags can be used.

**Comment by niruiyu** [comment]
@mdkinney , I guess it's a "draft" patch aiming to do a CI check.
There will be a complete set of patches in one PR to enable standalone MM for X86.

**Comment by td36** [comment]
> Some of the commit messages have more than one Signed-off-by. There should only be one of those tags in each commit. If more than one author contributed to the change, then additional Co-authored-by tags can be used.

Thanks for the comments Mike. It's a draft PR for CI check and internal review.

File: StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.c [hunk 1]
```diff
@@ -25,6 +25,203 @@ EFI_MM_COMMUNICATION_PROTOCOL  mMmCommunication = {
...
+// Data structure used to declare a table of protocol notifications and event
+// notifications required by the Standalone Mm environment
+//
+typedef struct {
+  BOOLEAN             Protocol;
+  BOOLEAN             CloseOnLock;
+  EFI_GUID            *Guid;
+  EFI_EVENT_NOTIFY    NotifyFunction;
+  VOID                *NotifyContext;
+  EFI_TPL             NotifyTpl;
```
Comment (niruiyu) [inline StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.c:42]: Remove this field as all are using TPL_CALLBACK

File: StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.c [hunk 2]
```diff
@@ -25,6 +25,203 @@ EFI_MM_COMMUNICATION_PROTOCOL  mMmCommunication = {
...
+  { FALSE, TRUE,  &gEfiEndOfDxeEventGroupGuid,        MmGuidedEventNotify,          &gEfiEndOfDxeEventGroupGuid,        TPL_CALLBACK, NULL },
+  //
+  // Declare event notification on EndOfDxe event.  This is used to set EndOfDxe event signaled flag.
+  //
+  { FALSE, TRUE,  &gEfiEndOfDxeEventGroupGuid,        MmEndOfDxeEventNotify,        &gEfiEndOfDxeEventGroupGuid,        TPL_CALLBACK, NULL },
+  //
+  // Declare event notification on Ready To Boot Event Group.  This is an extra event notification that is
+  // used to make sure SMRAM is locked before any boot options are processed.
+  //
+  { FALSE, TRUE,  &gEfiEventReadyToBootGuid,          MmReadyToLockEventNotify,     &gEfiEventReadyToBootGuid,          TPL_CALLBACK, NULL },
```
Comment (niruiyu) [inline StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.c:70]: Do you think that moving this to just below "gEfiDxeSmmReadyToLockProtocolGuid" entry is more readable?

File: StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.c [hunk 3]
```diff
@@ -25,6 +25,203 @@ EFI_MM_COMMUNICATION_PROTOCOL  mMmCommunication = {
...
+EFI_SMM_ACCESS2_PROTOCOL   *mSmmAccess;
+BOOLEAN                    mSmmLocked = FALSE;
+BOOLEAN                    mEndOfDxe  = FALSE;
+
+//
+// Data structure used to declare a table of protocol notifications and event
+// notifications required by the Standalone Mm environment
+//
+typedef struct {
+  BOOLEAN             Protocol;
```
Comment (niruiyu) [inline StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.c:37]: enum {  EventNotify, ProtocolNotify };

**Review by niruiyu — CHANGES_REQUESTED** [review]

File: StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.c (hunk 3 above)
Comment (td36) [reply StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.c]: Added a new enum NOTIFICATION_TYPE.

File: StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.c (hunk 2 above)
Comment (td36) [reply StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.c]: Moved it to just below "gEfiDxeSmmReadyToLockProtocolGuid" entry

File: StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.c (hunk 1 above)
Comment (td36) [reply StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.c]: Removed in latest code change

File: StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.c [hunk 4]
```diff
@@ -25,6 +25,189 @@ EFI_MM_COMMUNICATION_PROTOCOL  mMmCommunication = {
...
+//
+// Table of Protocol notification and GUIDed Event notifications that the Standalone Mm requires
+//
+MM_EVENT_NOTIFICATION  mMmEvents[] = {
+  //
+  // Declare protocol notification on DxeSmmReadyToLock protocols.  When this notification is established,
+  // the associated event is immediately signalled, so the notification function will be executed and the
+  // DXE SMM Ready To Lock Protocol will be found if it is already in the handle database.
+  //
+  { ProtocolNotify, TRUE,  &gEfiDxeSmmReadyToLockProtocolGuid, MmReadyToLockEventNotify,     &gEfiDxeSmmReadyToLockProtocolGuid, NULL },
```
Comment (xuweiintel) [inline StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.c:41]: Better to use gEfiDxeMmReadyToLockProtocolGuid, which is consistent with the one in StandaloneMmCore

Comment (td36) [reply StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.c]: Updated

File: StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.h [hunk 5]
```diff
@@ -91,4 +113,64 @@ MmCommunicate (
...
+  This is a notification function registered on EVT_SIGNAL_VIRTUAL_ADDRESS_CHANGE event.
+  It convers pointer to new virtual address.
+
+  @param[in]  Event        Event whose notification function is being invoked.
+  @param[in]  Context      Pointer to the notification function's context.
+
+**/
+VOID
+EFIAPI
+MmVariableAddressChangeEvent (
```
Comment (niruiyu) [inline StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.h:171]: MmVirtualAddressChange, not MmVariableAddressChange

Comment (td36) [reply StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.h:171]: Updated

**Review by chethangowda6 — APPROVED** [review]

File: StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.h (hunk 5 above)
Comment (niruiyu) [reply StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.h:171]: it's still the old name. you only update the function name in array.

Comment (td36) [reply StandaloneMmPkg/Drivers/MmCommunicationDxe/MmCommunicationDxe.h:171]: I just found the wrong name in headfile from CI check and was about to update

**Pull request closed without merging**